beacon-chain/db/filesystem/layout_by_epoch.go

```diff
@@ -0,0 +1,193 @@
+package filesystem
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v5/config/params"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/primitives"
+	"github.com/spf13/afero"
+)
+
+type periodicEpochLayout struct {
+	fs     afero.Fs
+	cache  *blobStorageCache
+	pruner *blobPruner
+}
+
+var _ fsLayout = &periodicEpochLayout{}
+
+func newPeriodicEpochLayout(fs afero.Fs, cache *blobStorageCache, pruner *blobPruner) fsLayout {
+	l := &periodicEpochLayout{fs: fs, cache: cache, pruner: pruner}
+	return l
+}
+
+func (l *periodicEpochLayout) notify(ident blobIdent) error {
+	if err := l.cache.ensure(ident); err != nil {
+		return err
+	}
+	l.pruner.notify(ident.epoch, l)
+	return nil
+}
+
+// If before == 0, it won't be used as a filter and all idents will be returned.
+func (l *periodicEpochLayout) iterateIdents(before primitives.Epoch) (*identIterator, error) {
+	_, err := l.fs.Stat(periodicEpochBaseDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return &identIterator{eof: true}, nil // The directory is non-existent, which is fine; stop iteration.
+		}
+		return nil, errors.Wrapf(err, "error reading path %s", periodicEpochBaseDir)
+	}
+	// iterate root, which should have directories named by "period"
+	entries, err := listDir(l.fs, periodicEpochBaseDir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list %s", periodicEpochBaseDir)
+	}
+
+	return &identIterator{
+		fs:   l.fs,
+		path: periodicEpochBaseDir,
+		levels: []layoutLevel{
```

Comment:
OK I understand now.
It was not clear to me at all that:
- the first item of the slice corresponds to the `period` level of the path
- the seconde one corresponds to the `epoch` level of the path
- ...


Maybe it's just me.
But I guess adding a comment explaining this could give a nice tip to a potential future reader.

(I had to run tests in debug mode to understand actually.)

Reply:
Out of curiosity, did you see the commentary on `identIterator` about how the levels work? I assumed readers would look at that, and that the naming of the populate/filter funcs in the levels would be intuitive. But I agree that adding comments on the levels here is helpful too!

Reply:
Yes I read it. But I was still confused about the role of `levels`. By again, it's maybe just me.